xcdat/temporal_avg.py

```diff
@@ -0,0 +1,913 @@
+"""Module containing temporal averaging functions."""
+
+from itertools import chain
+from typing import Dict, List, Optional, Tuple, Union
+
+import cf_xarray  # noqa: F401
+import numpy as np
+import pandas as pd
+import xarray as xr
+from typing_extensions import Literal, TypedDict, get_args
+
+from xcdat import bounds  # noqa: F401
+from xcdat.dataset import get_data_var
+from xcdat.logger import setup_custom_logger
+from xcdat.utils import bool_str_to_bool
+
+logger = setup_custom_logger(__name__)
+
+# Type alias for supported time averaging modes.
+Mode = Literal["climatology", "timeseries", "departure"]
+MODES = get_args(Mode)
+
+# Type alias for supported grouping frequencies.
+Frequency = Literal["hour", "day", "month", "season", "year"]
+#: Tuple of supported grouping frequencies.
+FREQUENCIES = get_args(Frequency)
+
+# Configuration specific to the "season" frequency.
+SeasonConfig = TypedDict(
+    "SeasonConfig",
+    {
+        "winter_season": Literal["DJF", "JFD"],
```

Comment:
I think we should make this `boreal_winter_season`

Reply:
@pochedls Is this the technical term you found for this key?

Reply:
It seems like we should base this off [meteorological winter](https://en.wikipedia.org/wiki/Season#Meteorological), which unfortunately is different in each hemisphere, which is why I appended "boreal." I think it is okay if this is a little long, since we can define the default value (DJF) which I assume won't change that much (I guess if you are passing in a climatology you'd have to change the default value?).

Reply:
`boreal_winter_season` works since there doesn't seem to be a more generalized term for that season across hemispheres.

> (I guess if you are passing in a climatology you'd have to change the default value?).
If the climatology is calculated using xcdat, this attribute is stored in the variable DataArray for reference.
Otherwise, the user would have to change the value if it should be "JFD".

Reply:
boreal_winter_season is accurate. or `NH_winter`, which is short for `Northern_Hemisphere_Winter`,  how about `dec_mod` shorter?

Reply:
Thanks @chengzhuzhang. `dec_mode` is a good alternative so that the key doesn't have to be associated with a "winter" season.